### PR TITLE
test: disable artifacts cache with composefs

### PR DESCRIPTION
**Pull request by giuseppe**
layers restored from a tarball won't be converted to composefs (for now at least), so disable the cache when using composefs.

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: test/utils/matchers.go
Comment (edsantiago) [inline test/utils/matchers.go:110]: Yikes. Please add an enormous FIXME comment here explaining why the NO_TEST_CACHE exclusion is necessary and when it can be removed. I'm really really uncomfortable with this, is there any other way to be more selective about when/where warnings are silenced?

Comment (giuseppe) [reply test/utils/matchers.go:110]: Many tests fail right now with `NO_TEST_CACHE` because they perform an image pull.

This one was the easiest fix, the alternative would be to filter out any output generated by a pull, or pass `-q` to each command that could cause a pull.

Comment (edsantiago) [reply test/utils/matchers.go:112]: I thought I already had added `-q` to any tests that might do a pull. Do you have a log handy, showing sample failures?

Also, in the other diff, please never say "for now". Please always include exact datestamps and, if possible, issue numbers, so a future reader can think "hmmm this should be fixed by now".

**Comment by edsantiago** [comment]
Was this supposed to fix the lxgetattr flake? It does not seem to: https://api.cirrus-ci.com/v1/artifact/task/5139484897968128/html/int-podman-rawhide-root-host-sqlite.log.html

**Comment by giuseppe** [comment]
> Was this supposed to fix the lxgetattr flake? It does not seem to: https://api.cirrus-ci.com/v1/artifact/task/5139484897968128/html/int-podman-rawhide-root-host-sqlite.log.html

no, I am still trying to find a reproducer.

File: test/utils/matchers.go
Comment (giuseppe) [reply test/utils/matchers.go]: I've added a "revert" patch to the current PR so we can get a list of the failing tests

Comment (giuseppe) [reply test/utils/matchers.go]: I was wrong, it seems we don't need this change

**Comment by giuseppe** [comment]
@edsantiago here are some of the failing tests: https://github.com/containers/podman/pull/23274/checks?check_run_id=27476648865

**Comment by edsantiago** [comment]
OIC. Now I see what `NO_TEST_CACHE` does. That... is not a solution. We can't run CI like that. If that's the only way to get CI to pass, let's just disable composefs.

**Comment by giuseppe** [comment]
tests seem to pass now, but we need these three fixes in storage:

- https://github.com/containers/storage/pull/2031
- https://github.com/containers/storage/pull/2036
- https://github.com/containers/storage/pull/2037

@edsantiago If you'd like to test with the potential fix for the flake,  I've prepared a branch that includes all the fixes above: https://github.com/giuseppe/storage/tree/test-composefs-fixes-for-podman

**Comment by giuseppe** [comment]
images copied to the RW store with `AddImageToRWStore` won't be stored in composefs mode.  This may change in future, if we automatically convert images not pulled from a registry.  For now I think it is desiderable behavior so we can test both variants: images stored in composefs mode in the additional store, and images stored in the traditional format when imported.

**Comment by giuseppe** [comment]
ready for review.

@edsantiago PTAL

**Comment by edsantiago** [comment]
Thanks @giuseppe. This is a lot of scariness to review, I'm really uncomfortable with some of the changes, so I've just been hammering away in #17831. I just saw [this failure](https://api.cirrus-ci.com/v1/artifact/task/6664205737132032/html/int-remote-rawhide-root-host-sqlite.log.html#t--Podman-commit-podman-commit-adds-exposed-ports--1). Is it composefs?
```
# podman-remote [options] run --name testcon -p 8585:80 quay.io/libpod/alpine:latest true
# podman-remote [options] commit -q testcon newimage
Error: copying layers and metadata for container "cecc9c90f4107cf89808b8234df98e173e1bf052a006665f460a1e86f8d070b8": initializing source containers-storage:testcon: extracting layer "1745071114b3933b11b73a9f099d2b82d4e80262c1211b1b6e7f388466ebdf3c": invalid argument
```

**Comment by giuseppe** [comment]
> Thanks @giuseppe. This is a lot of scariness to review, I'm really uncomfortable with some of the changes, so I've just been hammering away in #17831. I just saw [this failure](https://api.cirrus-ci.com/v1/artifact/task/6664205737132032/html/int-remote-rawhide-root-host-sqlite.log.html#t--Podman-commit-podman-commit-adds-exposed-ports--1). Is it composefs?

difficult to say, commit doesn't create a composefs image but it still access the layers stored in composefs from the additional image store.  I've not seen that error before

**Comment by edsantiago** [comment]
Another one, rawhide, [on commit](https://api.cirrus-ci.com/v1/artifact/task/6461293362675712/html/int-remote-rawhide-root-host-sqlite.log.html#t--Podman-commit-podman-commit-container-check-env-variables--1). I'm not saying this is composefs, but this is a flake I've never seen before today.

Also another new flake, but I'll file a new issue for that.

File: contrib/cirrus/setup_environment.sh
Comment (edsantiago) [inline contrib/cirrus/setup_environment.sh:58]: Is this a "will never be stored correctly", or "there's a bug that we need to fix"? Future maintainers will need to know this. (And, if the latter, will want an issue number).

Also, I suggest renaming the variable to `CI_NO_TEST_ARTIFACTS` or something with a `CI_` prefix. A `CI_` prefix is treated specially, such that it gets propagated to all test environments (like rootless). I know composefs is root-only, but maybe some day that will change, or maybe this NO_ARTIFACT thing will get reused for a different purpose some day, and whoever deals with that may spend a lot of time wondering why rootless isn't working. Naming it properly today will be a silent gift to that person.

**Review by edsantiago — COMMENTED** [review]
This makes me super uncomfortable. It seems like a kludgy workaround for bugs in composefs, and I worry that some day someone else may use it to work around other future bugs. And this only works because of the new local registry; prior to that it would pull from the net, which would be slow and flaky.

The commit stack is also deeply troubling: I had a lot of trouble reviewing commit-by-commit because work in one is undone in the next. If this is to be approved, may I encourage you to give some thought to cleaning up the individual commits?

I will leave it to others to approve-merge this. Right now I don't feel comfortable doing so.

**Comment by giuseppe** [comment]
> This makes me super uncomfortable. It seems like a kludgy workaround for bugs in composefs, and I worry that some day someone else may use it to work around other future bugs. And this only works because of the new local registry; prior to that it would pull from the net, which would be slow and flaky.

the images are pulled only once, and stored in the additional image store so it would still work pulling them from the network.

It is a workaround for composefs, since it supports only pulling from a registry at the moment.  We might need to support pulling from other sources to enable "conversion" of images.  That was not initially needed for partial pulls, so it is something we still need to implement.

File: contrib/cirrus/setup_environment.sh
Comment (giuseppe) [reply contrib/cirrus/setup_environment.sh:58]: I used a named similar to the existing `NO_TEST_CACHE`.  Renamed to `CI_NO_TEST_ARTIFACTS`.

**Comment by giuseppe** [comment]
> The commit stack is also deeply troubling: I had a lot of trouble reviewing commit-by-commit because work in one is undone in the next. If this is to be approved, may I encourage you to give some thought to cleaning up the individual commits?

I've not noticed that these pieces were left in the second commit.  Dropped them

**Comment by martinpitt** [comment]
Wrt. the `chpasswd` [breakage in rawhide](https://artifacts.dev.testing-farm.io/2eee4823-901e-44c9-9ae3-f8b58b4bd341/): This was fallout from an intermediate bad shadow-utils upload in rawhide. See  https://github.com/stratis-storage/stratisd/pull/3623#issuecomment-2246827933 ff. for details.

TL/DR: Please retry the failed tests here and in #23275 and other recent PRs. Thanks!

**Comment by giuseppe** [comment]
@containers/podman-maintainers PTAL

File: test/e2e/e2e.test
Comment (Luap99) [inline test/e2e/e2e.test:1]: this should not be commited

File: contrib/cirrus/setup_environment.sh
Comment (Luap99) [reply contrib/cirrus/setup_environment.sh:58]: I think the question is still open here, is this temporary or permanent?

File: contrib/cirrus/setup_environment.sh
Comment (Luap99) [inline contrib/cirrus/setup_environment.sh:59]: Why do we need another env var at all? If I wanted to run e2e locally now I need to set 

```
export STORAGE_FS=overlayfs
export STORAGE_OPTIONS_OVERLAY='overlay.use_composefs=true --pull-option=enable_partial_images=true --pull-option=convert_images=true'
````
and now yet another...
The e2e tests should be smart enough to know they are already configured to use composefs and then you can skip the cache. I don't see why anyone would want need this option without using composefs

Comment (giuseppe) [reply contrib/cirrus/setup_environment.sh]: dropped the additional environment variable and just look if composefs is enabled

**Comment by edsantiago** [comment]
New flake (as in, one I've never seen before, and is happening on [rawhide root](https://api.cirrus-ci.com/v1/artifact/task/5036223750733824/html/int-podman-rawhide-root-host-sqlite.log.html#t--Podman-checkpoint-podman-checkpoint-create-image-with-running-container--1):
```
# podman [options] --pull-option=enable_partial_images=true --pull-option=convert_images=true container checkpoint --create-image alpine-checkpoint-bmjmtt --keep e3793c9b35baf7d6a62a976fb4da4e4fc139c535edd159cd6edcf96a8b4e1d02
Error: exporting root file-system diff for "e3793c9b35baf7d6a62a976fb4da4e4fc139c535edd159cd6edcf96a8b4e1d02": invalid argument
```
I'm blaming composefs. Should I open a new issue for it?

**Comment by giuseppe** [comment]
yes, I think it is the same root issue as https://github.com/containers/podman/pull/23274#issuecomment-2243363751

**Comment by edsantiago** [comment]
https://github.com/containers/storage/issues/2042

**Comment by Luap99** [comment]
Should we then not merge for now given there is a (new?) flake? Or does this fix enough other flakes/problems that this should be merged regardless?

**Comment by edsantiago** [comment]
Speaking as the person running the no-flake-retry tests, this is a big improvement over the current state of composefs in podman. With this PR in #17831, I see fewer flakes. Still see flakes, and I fully expect there to be more, but as-is right now composefs is unusable. Incremental progress is ok with me.

**Comment by edsantiago** [comment]
That said, if the EINVAL flake is an easy one to find/fix, it'd sure be nice to whack it

**Comment by giuseppe** [comment]
without this PR we are not creating the files in the additional image store in the composefs format

**Comment by mheon** [comment]
LGTM

**Comment by giuseppe** [comment]
let's wait for https://github.com/containers/storage/pull/2043 and https://github.com/containers/storage/pull/2044 before merging.  Hopefully they can help to troubleshoot the "invalid argument" flake

**Comment by giuseppe** [comment]
patches for c/storage included, ready to merge

**Review by Luap99 — APPROVED** [review]
/lgtm

**Pull request merged**
